File: app/javascript/channels/kitchen_channel.js
import consumer from "./consumer"

export default class KitchenChannel {
  constructor(restaurantId, callbacks = {}) {
    this.restaurantId = restaurantId
    this.callbacks = callbacks
    this.subscription = null
  }
  
  connect() {
    this.subscription = consumer.subscriptions.create(
      { channel: "KitchenChannel", restaurant_id: this.restaurantId },
      {
        connected: () => {
          console.log(`Connected to kitchen channel for restaurant ${this.restaurantId}`)
          if (this.callbacks.onConnected) this.callbacks.onConnected()
        },

        disconnected: () => {
          console.log("Disconnected from kitchen channel")
          if (this.callbacks.onDisconnected) this.callbacks.onDisconnected()
        },

        received: (data) => {
          console.log("Kitchen update received:", data)
          this.handleMessage(data)
        }
      }
    )
  }
  
  disconnect() {
    if (this.subscription) {
      this.subscription.unsubscribe()
    }
  }
  
  handleMessage(data) {
    switch(data.event) {
      case 'new_order':
        if (this.callbacks.onNewOrder) this.callbacks.onNewOrder(data.order)
        break
      case 'status_change':
        if (this.callbacks.onStatusChange) this.callbacks.onStatusChange(data)
        break
      case 'inventory_alert':
        if (this.callbacks.onInventoryAlert) this.callbacks.onInventoryAlert(data)
        break
      case 'staff_assignment':
        if (this.callbacks.onStaffAssignment) this.callbacks.onStaffAssignment(data)
        break
      case 'queue_update':
        if (this.callbacks.onQueueUpdate) this.callbacks.onQueueUpdate(data)
        break
      case 'metrics_update':
        if (this.callbacks.onMetricsUpdate) this.callbacks.onMetricsUpdate(data)
        break
    }
  }
  
  updateStatus(orderId, newStatus) {
    if (this.subscription) {
      this.subscription.send({
        action: 'update_status',
        order_id: orderId,
        new_status: newStatus
      })
    }
  }
  
  assignStaff(orderId, staffId) {
    if (this.subscription) {
      this.subscription.send({
        action: 'assign_staff',
        order_id: orderId,
        staff_id: staffId
      })
    }
  }
}

File: app/channels/kitchen_channel.rb
class KitchenChannel < ApplicationCable::Channel
  def subscribed
    restaurant_id = params[:restaurant_id]
    return reject unless restaurant_id

    stream_from "kitchen_#{restaurant_id}"

    # Track presence
    return unless current_user

    PresenceService.user_online(
      current_user,
      connection.connection_identifier,
      resource_type: 'Restaurant',
      resource_id: restaurant_id,
    )
  end

  def unsubscribed
    # Mark user as offline
    return unless current_user

    PresenceService.user_offline(
      current_user,
      connection.connection_identifier,
    )
  end

  def receive(data)
    # Handle incoming messages (e.g., status updates, assignments)
    case data['action']
    when 'update_status'
      handle_status_update(data)
    when 'assign_staff'
      handle_staff_assignment(data)
    end
  end

  private

  def handle_status_update(data)
    order = Ordr.find_by(id: data['order_id'])
    return unless order

    old_status = order.status
    order.update(status: data['new_status'])

    KitchenBroadcastService.broadcast_status_change(
      order,
      old_status,
      data['new_status'],
    )
  end

  def handle_staff_assignment(data)
    order = Ordr.find_by(id: data['order_id'])
    staff = User.find_by(id: data['staff_id'])
    return unless order && staff

    KitchenBroadcastService.broadcast_staff_assignment(order, staff)
  end
end
